File: Data Structures & Algorithms/subarray-sum-equals-k/submission-0.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int n = nums.length;
        
        // pass1: build prefix Sum
        int[] prefix = new int[n + 1];
        for(int i = 0; i < n; i++){
            prefix[i + 1] = nums[i] + prefix[i];
        }

        // pass2: count how many prefix[i] == prefix[j] - k
        Map<Integer, Integer> map = new HashMap<>();
        int res = 0;
        for(int j = 0; j <= n; j++){
            // how many earlier prefix Sum equal to prefix[j] - k
            res += map.getOrDefault(prefix[j] - k, 0);
            // record current prefix sum
            map.put(prefix[j], map.getOrDefault(prefix[j], 0) + 1);
        }
        return res;
    }
}

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = -1
        r = len(nums) - 1 # (-1, n-1)
        while l + 1 < r:
            mid = (l + r) // 2
            x = nums[mid]
            if target > nums[-1] >= x: # target 在第一段，x在第二段
                r = mid
            elif x > nums[-1] >= target: # target在第二段，x在第一段
                l = mid
            elif x >= target: # x和target在同一段
                r = mid
            else:
                l = mid

        return r if nums[r] == target else -1